compiler/rustc_typeck/src/check/inherited.rs

```diff
@@ -89,7 +93,29 @@ impl<'tcx> Inherited<'_, 'tcx> {
             infcx: tcx
                 .infer_ctxt()
                 .ignoring_regions()
-                .with_fresh_in_progress_typeck_results(hir_owner),
+                .with_fresh_in_progress_typeck_results(hir_owner)
+                .with_normalize_fn_sig_for_diagnostic(Lrc::new(move |infcx, fn_sig| {
+                    if fn_sig.has_escaping_bound_vars() {
+                        return fn_sig;
+                    }
+                    infcx.probe(|_| {
+                        let traits::Normalized { value: normalized_fn_sig, obligations } =
+                            traits::normalize(
+                                &mut traits::SelectionContext::new(infcx),
+                                // FIXME(compiler-errors): This is probably not the right param-env...
+                                infcx.tcx.param_env(def_id),
+                                ObligationCause::dummy(),
+                                fn_sig,
+                            );
+                        let mut fulfillment_ctxt = FulfillmentContext::new_in_snapshot();
+                        fulfillment_ctxt.register_predicate_obligations(infcx, obligations);
+                        if fulfillment_ctxt.select_all_or_error(infcx).is_empty() {
+                            infcx.resolve_vars_if_possible(normalized_fn_sig)
```

Comment:
we potentially leak inference variables here. It would be easiest to only return here if `normalized_fn_sig` doesn't contain inference vars.